transaction/std/src/transaction_builder.rs

```diff
@@ -139,7 +142,72 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
     ///   signature for an input.
     pub fn add_input(&mut self, input_credentials: InputCredentials) {
-        self.input_credentials.push(input_credentials);
+        self.input_materials
+            .push(InputMaterials::Signable(input_credentials));
+    }
+
+    /// Add a pre-signed Input to the transaction, also fulfilling any
+    /// requirements imposed by the signed rules, so that our transaction
+    /// will be valid.
+    ///
+    /// Note: Before adding a signed_contingent_input, you probably want to:
+    /// * validate it (call .validate())
+    /// * check if key image appreared already (call .key_image())
+    /// * provide merkle proofs of membership for each ring member (see
+    ///   .tx_out_global_indices)
+    ///
+    /// # Arguments
+    /// * `signed_contingent_input` - The pre-signed input we are adding
+    pub fn add_presigned_input(
+        &mut self,
+        sci: SignedContingentInput,
+    ) -> Result<(), SignedContingentInputError> {
+        if let Some(rules) = sci.tx_in.input_rules.as_ref() {
+            // Enforce all rules so that our transaction will be valid
+            if rules.required_outputs.len() != sci.required_output_amounts.len() {
+                return Err(SignedContingentInputError::WrongNumberOfRequiredOutputAmounts);
+            }
+            // 1. Required outputs
+            for (required_output, unmasked_amount) in rules
+                .required_outputs
+                .iter()
+                .zip(sci.required_output_amounts.iter())
+            {
+                // Check if the required output is already there
+                if !self
+                    .outputs_and_secrets
+                    .iter()
+                    .any(|(output, _sec)| output == required_output)
+                {
+                    // If not, add it
+                    self.outputs_and_secrets
+                        .push((required_output.clone(), unmasked_amount.clone().into()));
+                }
+            }
+            // 2. Max tombstone block
+            if rules.max_tombstone_block != 0 {
+                self.impose_tombstone_block_limit(rules.max_tombstone_block);
+            }
+        }
+
+        self.add_presigned_input_raw(sci);
+        Ok(())
+    }
+
+    /// Add a pre-signed Input to the transaction, without also fulfilling
+    /// any of its rules. You will have to add any required outputs, adjust
+    /// tombstone block, etc., for the transaction to be valid.
+    ///
+    /// Note: Before adding a signed_contingent_input, you probably want to:
+    /// * validate it (call .validate())
+    /// * check if key image appreared already (call .key_image())
+    /// * provide merkle proofs of membership for each ring member (see
+    ///   .tx_out_global_indices)
+    ///
+    /// # Arguments
+    /// * `signed_contingent_input` - The pre-signed input we are adding
+    pub fn add_presigned_input_raw(&mut self, sci: SignedContingentInput) {
```

Comment:
Does this need to be `pub`?

Reply:
So, potentially in this revision it could just be `pub(crate)`.

But in the future it may have to be `pub`: for instance, one feature that was requested was to have rules that could be satisfied in more than one way, for instance like "I will trade my MOB for 100 of this or 200 of that". In that case, the `add_presigned_input` approach won't really be able to work because it won't know which way you want to satisfy the rule. Right now I think the simplest thing to do for that would be to say that, you can add it using the `_raw` version and then have like an `add_output_raw` so that you add the outputs that satisfy it the way you want to.

Brian was separately requesting an `add_output_raw` for use in the hardware wallet project. I'm not sure if we still need that but it's at least conceivable that it's useful in cases like that.